### PR TITLE
Experiment: Add the Link component

### PR DESCRIPTION
**Pull request by diegohaz**
## Motivation

Related to #7112.

> [!IMPORTANT]
> This pull request is an implementation experiment. It is not intended to merge, and #7112 stays open.

Native links support modifier clicks, new tabs, context menus, copying, dragging, mobile previews, and other browser behaviors that button-driven router navigation cannot reproduce. HTML anchors have no native `disabled` state. Removing `href` stops navigation, but it also removes implicit link semantics.

This experiment puts the proposed `Link` API into practice so we can evaluate the design against native anchors, server rendering, React Router, and the Next.js App Router.

## Solution

This pull request adds experimental `Link` and `useLink` exports at `@ariakit/react-components/link/link`. It does not add an `@ariakit/react` re-export, a reference page, `LinkLabel`, or `LinkDescription`.

`Link` owns its native destination. When disabled, it omits `href`, `target`, `download`, `ping`, `rel`, `hrefLang`, `referrerPolicy`, `type`, and anchor `itemProp` during render. The server HTML already has `role="link"`, `aria-disabled="true"`, the correct tab stop, and no native `disabled` attribute.

```tsx
import { Link } from "@ariakit/react-components/link/link";

<Link href={nextPage} disabled={!nextPage}>
  Next page
</Link>;
```

The patch also teaches `Focusable` about disabled anchors. An href-less disabled anchor keeps link semantics, and `accessibleWhenDisabled` can keep it in the tab order. This affects stable disabled-anchor behavior, including an explicit redundant `role="link"` on disabled `a[href]` elements.

Development warnings report non-anchor render elements and render-owned destinations that `Link` cannot remove. A downstream production build removes the warning-only code and strings.

### Native link

Pass the destination directly to `Link`. This keeps one anchor mounted when the state changes.

```tsx
<Link href="/reports" disabled={reportCount === 0} accessibleWhenDisabled>
  Reports
</Link>
```

### React Router DOM 6.30.4

React Router exposes the low-level pieces needed for a minimal stable-node adapter. This is not a drop-in replacement for every React Router `Link` option.

```tsx
function RouterLink({ to, onClick, target, ...props }) {
  const href = useHref(to);
  const handleClick = useLinkClickHandler(to, { target });

  return (
    <Link
      {...props}
      href={href}
      target={target}
      onClick={(event) => {
        onClick?.(event);
        if (event.defaultPrevented) return;
        handleClick(event);
      }}
    />
  );
}
```

### Next.js 16.2.12 App Router

`NextLink` owns and requires its `href`. The safe adapter must resolve the disabled state before it selects the render element.

```tsx
const linkDisabled =
  focusable !== false &&
  (disabled || ariaDisabled === true || ariaDisabled === "true");

<Link
  disabled={disabled}
  aria-disabled={ariaDisabled}
  focusable={focusable}
  render={linkDisabled ? <a /> : <NextLink href={href} />}
/>;
```

This conditional element is server-correct, but React remounts the anchor at the boundary. Focus is lost when `accessibleWhenDisabled` keeps the new disabled anchor reachable.

## Preview sandboxes

| Use case | Uploaded preview |
| --- | --- |
| Native `Link` contract, SSR, tab order, composition, and attribute restoration | [Open the native Link sandbox](https://link-7112-experiment-ariakit-preview.ariakit.workers.dev/react/previews/link-disabled/) |
| Stable React Router adapter and consumer event control | [Open the React Router sandbox](https://link-7112-experiment-ariakit-preview.ariakit.workers.dev/react/previews/link-react-router/) |
| Next App Router conditional adapter and focus-loss limitation | [Open the Next.js sandbox](https://link-7112-experiment-ariakit-nextjs.ariakit.workers.dev/link-nextjs-app-router) |

### Native `Link`

[Open the native `Link` sandbox](https://link-7112-experiment-ariakit-preview.ariakit.workers.dev/react/previews/link-disabled/)

The page-3 boundary renders a disabled Next anchor without a destination, while the surrounding pagination remains native:

Native Link pagination sandbox at page 3 with the Next link disabled

The same focused anchor withholds and restores its destination without losing node identity or focus:

[Watch the native `Link` focus and destination recording](https://github.com/user-attachments/assets/b7c93803-1ff9-4834-901b-03fa8144495f)

### React Router

[Open the React Router sandbox](https://link-7112-experiment-ariakit-preview.ariakit.workers.dev/react/previews/link-react-router/)

The stable adapter keeps the Previous anchor focused after the route changes:

React Router Link sandbox with the Previous page anchor focused after navigation

The recording also shows that consumer `preventDefault()` wins before router navigation:

[Watch the React Router stable-node and event-guard recording](https://github.com/user-attachments/assets/9c69dd96-309a-4c1d-9de1-daed87fa3c05)

### Next.js App Router

[Open the Next.js App Router sandbox](https://link-7112-experiment-ariakit-nextjs.ariakit.workers.dev/link-nextjs-app-router)

The page-3 route produces the correct destination-free disabled anchor on the server:

Next.js App Router Link sandbox at page 3 with destination-free disabled output

Keyboard navigation across the `NextLink` to anchor render boundary demonstrates the focus-loss limitation:

[Watch the Next.js render-boundary focus recording](https://github.com/user-attachments/assets/f61ff18a-c495-413e-8502-82f685093e3a)

## What worked well

- Link-owned destination props make disabled output correct during render, SSR, hydration, and JavaScript-disabled use. Attributes restore declaratively when the link becomes enabled.
- The native adapter keeps one anchor mounted, so node identity and focus survive enabled and disabled transitions.
- Existing `Focusable` behavior supplies disabled interaction guards, pointer behavior, ARIA state, and optional keyboard reachability.
- Outer components keep their more specific role. The sandbox verifies a disabled `MenuItem` rendered with `Link`.
- React Router DOM 6.30.4 exposes enough low-level hooks for a stable native anchor adapter. The adapter can also call the consumer first and respect `preventDefault()`.
- Development warnings catch the two important composition mistakes, and a downstream production bundle removes them.

## What did not work well

- Render-element props win Ariakit composition. `Link` cannot remove an `href` owned or derived by `NextLink` or another render component. Before hydration, that destination is still active. After hydration, `NextLink` can also prefetch it. The warning is best-effort because it requires a forwarded ref and a mounted element.
- The safe Next adapter swaps `NextLink` for `<a>`. That produces correct server HTML but remounts the node and loses focus at the disabled boundary.
- The Next adapter must duplicate `Link` disabled-state resolution before choosing its render element. Checking only the boolean `disabled` prop leaves an ARIA-disabled link live on the server.
- `focusable={false}` intentionally turns off the inherited `Focusable` layer, including disabled resolution. Therefore, `<Link disabled focusable={false} href="/target">` stays live. This matches the proposed design and existing `Focusable` semantics, but it is a sharp edge. Decoupling destination removal from focusability would require an explicit API redesign for ARIA and tab-order behavior.
- Server correctness required initial capability hints for a statically supplied `<a />` render target. Function render callbacks and custom render components do not expose their intrinsic element type to `useFocusable` during server render, so they keep mount-time capability resolution. Solving that generally would require an explicit native-element hint or a render-pipeline redesign. Experimental `Link` remains server-correct for callback renders because it knows its anchor contract and owns the destination.
- Implementing the API exposed a broader stable disabled-anchor change in `Focusable` and components built on it. Disabled `a[href]` elements now receive an explicit `role="link"`; ARIA in HTML allows this but does not recommend the redundant role.
- HTML conformance also required withholding anchor microdata `itemProp`, which was not in the original destination-attribute list.
- React does not repair an initial server/client disabled-state hydration mismatch. Stale attributes can persist until a remount or a later value-changing render forces a DOM write.
- Safari on macOS can require Option+Tab to traverse all controls when full keyboard navigation is off. The sandbox and browser test must present this as a fallback because Safari can swap the Tab and Option+Tab behaviors in its Advanced settings; this platform policy is outside `Link`'s `tabIndex` contract.
- Real screen-reader announcement and Safari click-focus behavior still need manual research before any merge proposal.

## Implementation review reassessment

<details>
<summary>Cycle 3: Narrow stable disabled-anchor SSR support</summary>

**Assessment**

- Issue severity: Moderate. Ariakit does not currently ship this `Link` API, and consumers have a small manual workaround. However, incorrect disabled-link output can expose the wrong semantics to assistive technology or leave navigation available during server-only use.
- Expected frequency: Disabled pagination and unavailable navigation links are common. The repeated generic `Focusable` edge requires a callback or opaque custom render target, a disabled anchor, and interaction before hydration or without JavaScript. That narrower combination should be uncommon, although its impact can be substantial when it occurs.
- Supported scope: Experimental `Link` supports destinations that it owns, boolean and ARIA disabled states, destination-attribute omission and restoration, SSR, hydration, JavaScript-disabled output, `accessibleWhenDisabled`, outer-role composition, and callback render anchors that forward the supplied props. The stable `Focusable` improvement supports SSR capability seeding only for a statically supplied intrinsic `render={<a />}`. Opaque targets keep the existing mount-time capability resolution.
- Likely user impact: The primary `Link` path produces correct inert and accessible markup from the initial server response. Static disabled `Focusable` anchors also gain correct role and tab-order output before hydration. Unsupported generic callback anchors can still retain an ineffective `disabled` attribute or an active destination before capability resolution.
- Implementation and maintenance complexity: The runtime design remains localized and adds no new effect, context, render pass, or public capability-hint API. The public-contract cost is moderate because the `Focusable` behavior also changes stable packages and must remain precisely described. Most of the larger evidence footprint comes from the three styled sandboxes and their SSR, hydration, JavaScript-disabled, router, and browser coverage.

**Decision**

Keep the statically supplied intrinsic-anchor optimization and narrow the stable `Focusable` guarantee. Generic callback and custom-component intrinsic types remain runtime-resolved. `Link` keeps callback SSR support and targeted coverage because it owns the anchor contract and destination.

Intentionally unsupported: generic `Focusable` SSR inference for function renders and custom render components without an explicit native-element hint.

</details>

## Verification

Checks run on the final reviewed content:

- `pnpm lint-fix`
- focused Vitest: 3 files, 19 tests
- root TypeScript and Astro checks: 0 diagnostics
- Next.js route type generation and TypeScript
- `pnpm build-app-lite`: Next.js production build, React Compiler, OpenNext worker bundle, and Astro production preview
- affected Chrome sandbox tests: 11 tests in normal parallel mode
- corrected tab-order test in Chrome, Firefox, and Safari: 3 tests

Broader checks ran on the same implementation before review-only cleanup and the final edge-case additions:

- current React suite: 270 files, 1,930 tests
- React 18 suite: 198 files, 1,052 tests
- three-browser sandbox matrix: 18 tests
- reference tests: 13 tests
- package build, package-condition resolution, and `npm pack --dry-run` contents
- downstream production bundle warning-string exclusion
- reference-free website build: 450 pages

CI on this pull request is the broad final-snapshot verification.

## Remaining scope

`LinkLabel` and `LinkDescription` remain deferred as described in #7112. This experiment also does not add a stable root export or reference page. Screen-reader behavior in NVDA, JAWS, and VoiceOver remains a manual pre-merge gate for any future proposal.

### CODE REVIEW

File: packages/ariakit-react-components/src/focusable/focusable.tsx [hunk 1]
```diff
@@ -449,6 +462,7 @@ export const useFocusable = createHook<TagName, FocusableOptions>(
       focusable && !!(elementCapabilities & nativeTabbableMask);
     const supportsDisabled =
       focusable && !!(elementCapabilities & supportsDisabledMask);
+    const anchor = focusable && !!(elementCapabilities & anchorMask);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-components/src/focusable/focusable.tsx:475]: Fixed in [`e84621691`](https://github.com/ariakit/ariakit/commit/e8462169194e3b8d8e627a88d58e0e6137412212): statically rendered anchors now seed their capabilities during the initial render, so disabled link role and tab order are present in server HTML. The same commit adds render/hydration coverage and a JavaScript-disabled Chrome regression.

File: packages/ariakit-react-components/src/link/link.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  type: undefined,
+} satisfies AnchorHTMLAttributes<HTMLType>;
+
+/**
+ * Returns props to create a `Link` component.
+ */
+export const useLink = createHook<TagName, LinkOptions>(
+  function useLink(props) {
+    const { focusable = true } = props;
+    const disabled = focusable && disabledFromProps(props);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-components/src/link/link.tsx:40]: This is not an issue under the experiment's proposed contract. `focusable={false}` intentionally disables inherited `Focusable` behavior, including disabled resolution; #7112 specifies this exact case as retaining the live `href`, and the sandbox tests lock it. Decoupling destination removal from focusability would require a separate API design for the resulting ARIA and tab-order behavior. I added this sharp edge to the experiment report.

**Comment by diegohaz** [comment]
## Review feedback summary

**Fixed**

- [`e84621691`](https://github.com/ariakit/ariakit/commit/e8462169194e3b8d8e627a88d58e0e6137412212) ([reply](https://github.com/ariakit/ariakit/pull/7267#discussion_r3838979394)): statically rendered disabled anchors now expose link role and tab order in server HTML, with render/hydration and JavaScript-disabled browser coverage.

**Invalid under the proposed contract**

- `focusable={false}` ([reply](https://github.com/ariakit/ariakit/pull/7267#discussion_r3838979703)): #7112 explicitly defines this prop as disabling the inherited `Focusable` mechanism and retaining the live destination. The experiment report now records the sharp edge and the API redesign that decoupling would require.

File: packages/ariakit-react-components/src/focusable/focusable.tsx [hunk 3]
```diff
@@ -432,9 +452,12 @@ export const useFocusable = createHook<TagName, FocusableOptions>(
...
     // Track only the element capabilities that affect the returned props so
-    // resolving the default native element doesn't require another render.
-    const [elementCapabilities, setElementCapabilities] = useState(
-      defaultElementCapabilities,
+    // resolving the default native element doesn't require another render. A
+    // statically rendered anchor can expose disabled semantics on the server.
+    const [elementCapabilities, setElementCapabilities] = useState(() =>
+      isValidElement(props.render) && props.render.type === "a"
+        ? getElementCapabilities("a")
+        : defaultElementCapabilities,
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-components/src/focusable/focusable.tsx:460]: Declined as a non-goal for generic `Focusable` callback renders. The callback output is opaque until the composition layer invokes it, so server inference would require a new native-element hint or a render-pipeline contract outside #7112. I narrowed the stable changeset guarantee to statically supplied intrinsic anchors and recorded callback/custom-component targets as mount-resolved. Experimental `Link` remains server-correct for callback renders because it owns the anchor contract and destination; [`af23fd018`](https://github.com/ariakit/ariakit/commit/af23fd01859f5d69b5568a42b962c407d9bc09db) adds SSR, hydration, and JavaScript-disabled proof coverage.

**Comment by diegohaz** [comment]
## Review feedback summary

**Fixed**

- [`af23fd018`](https://github.com/ariakit/ariakit/commit/af23fd01859f5d69b5568a42b962c407d9bc09db): callback-rendered experimental `Link` anchors now have explicit SSR, hydration, and JavaScript-disabled contract coverage.
- [`af23fd018`](https://github.com/ariakit/ariakit/commit/af23fd01859f5d69b5568a42b962c407d9bc09db): the native disabled-anchor regression now repeats role, tab-order, and native-attribute assertions after hydration, addressing [review 5002827662](https://github.com/ariakit/ariakit/pull/7267#pullrequestreview-5002827662).

**Declined as a non-goal**

- Generic `Focusable` callback and opaque custom-render SSR inference ([reply](https://github.com/ariakit/ariakit/pull/7267#discussion_r3840406951)): the stable changeset is now limited to statically supplied intrinsic anchors. A general fix needs a native-element hint or render-pipeline contract outside #7112. Experimental `Link` remains server-correct for callback renders.
